primary/tests/integration_tests.rs

```diff
@@ -14,22 +14,183 @@ use test_utils::{
+        4,
+        actual_result
+            .iter()
+            .filter(|&r| matches!(r.retrieval_result, Some(types::RetrievalResult::Batch(_))))
+            .count()
+    );
+
+    // And 1 Error is returned
+    let errors: Vec<&CollectionRetrievalResult> = actual_result
+        .iter()
+        .filter(|&r| matches!(r.retrieval_result, Some(types::RetrievalResult::Error(_))))
+        .collect::<Vec<_>>();
+
+    assert_eq!(1, errors.len());
+
+    // And check missing collection id is correct
+    let actual_missing_collection = match errors[0].retrieval_result.as_ref().unwrap() {
+        types::RetrievalResult::Error(e) => e.id.as_ref(),
+        _ => panic!("Should never hit this branch."),
+    };
+
+    assert_eq!(
+        &CertificateDigestProto::from(missing_block),
+        actual_missing_collection.unwrap()
+    );
+}
+
+#[tokio::test]
+async fn test_remove_collections() {
+    // TODO: remove tracing before submitting code
+    let config = telemetry_subscribers::TelemetryConfig {
+        service_name: "test".into(),
+        ..Default::default()
+    };
+    #[allow(unused)]
+    let guard = telemetry_subscribers::init(config);
+
+    let parameters = Parameters {
+        batch_size: 200, // Two transactions.
+        ..Parameters::default()
+    };
+    let (store, mut collection_ids, _) = setup(parameters.clone()).await;
+
+    // Test gRPC server with client call
+    let mut client = connect_to_validator_client(parameters.clone()).await;
+
+    // Test remove no collections
+    let request = tonic::Request::new(Collections {
+        collection_ids: vec![],
+    });
+
+    let status = client.remove_collections(request).await.unwrap_err();
+
+    assert!(status
+        .message()
+        .contains("Attemped to remove no collections!"));
+
+    // Test remove 1 collection
+    let block_to_be_removed = collection_ids.remove(0);
+    let request = tonic::Request::new(Collections {
+        collection_ids: vec![block_to_be_removed.into()],
+    });
+    let response = client.remove_collections(request).await.unwrap();
+    let actual_result = response.into_inner();
+
+    assert_eq!(Empty {}, actual_result);
+
+    assert!(
+        store
+            .certificate_store
+            .read(block_to_be_removed)
+            .await
+            .unwrap()
+            .is_none(),
+        "Certificate shouldn't exist"
+    );
+
+    // Test remove remaining collections, one collection has its batches intentionally
+    // missing but it should not return any errors.
+    let request = tonic::Request::new(Collections {
+        collection_ids: collection_ids.iter().map(|&c_id| c_id.into()).collect(),
+    });
+    let response = client.remove_collections(request).await.unwrap();
+    let actual_result = response.into_inner();
+
+    assert_eq!(Empty {}, actual_result);
+    for &block_id in collection_ids.iter() {
+        assert!(
+            store
+                .certificate_store
+                .read(block_id)
```

Comment:
In case you are not aware you can also read stuff in bulk via the `read_all` method. Nothing wrong of course with this way as well

Example: https://github.com/MystenLabs/narwhal/blob/b7d2dfdd6394afa31cb910be0ff53efe7c6a5578/primary/src/helper.rs#L205

Reply:
I was not, thanks!